Object ref maps and RefMaps static utilities.

tinevez opened the pull request:
This PR adds a missing and required `ObjectRefMap`, that maps normal object to pool objects.

Also, all map related methods have been moved from `RefCollections` to `RefMaps` which break the API. After this PR is merged, some changes have to be made in `mastodon-graph` and `TrackMate3`.

tinevez committed:
ObjectRefHashMap: a map that maps plain objects to possibly reusable refs.

tinevez committed:
Junit tests for ObjectRefHashMap.

tinevez committed:
Specialized wrappers for heterogeneous maps.

We still want to expose the RefRefMap interface, and correctly
deal with create*Ref() methods in the cases where we have only
one of the two RefCollections that are pools.

tinevez committed:
Move map related methods from RefCollections to RefMaps.

This breaks the API.

tpietzsch commented:
I don't get the point of `RefObjectMap`. Where would this ever be required? I.e., when would you need to create a key ref?

tinevez commented:
You mean `ObjectRefMap`?

tpietzsch committed:
formatting

tpietzsch committed:
add javadoc

tpietzsch committed:
Minimal augmentation for ObjectRefHashMapTest

tpietzsch committed:
Revise RefObjectMap

 * create/releaseKeyRef() methods are not useful. removed.
 * specialize keySet() to return a RefSet

tpietzsch committed:
javadoc

tpietzsch committed:
Revise RefObjectMap

Specialize values() to return a RefCollection.
Ref-based implementations ObjectRefHashMap and RefRefHashMap more
specifically return RefPoolBackedRefCollection such that
RefCollections.createXXX() with values() argument will return proper
Ref-based implementations.

tpietzsch committed:
Move common parts of *MapWrappers to abstract base class

tpietzsch committed:
Move all RefRefMapWrappers into one class

Now:
RefRefMapWrapper.FromMap
RefRefMapWrapper.FromObjectRefMap
RefRefMapWrapper.FromRefObjectMap

tpietzsch committed:
Remove duplicate tryGetRefPool method

tpietzsch committed:
Rename "collection" parameters to "keyCollection" or "valueCollection". Clean up.
